File: packages/durable-streams/packages/client-java/src/main/java/com/durablestreams/model/Offset.java
package com.durablestreams.model;

import java.util.Objects;

/**
 * Opaque offset token identifying a position within a stream.
 *
 * <p><b>Comparison scope</b>: Offsets are only comparable within the SAME stream.
 * Comparing offsets from different streams is undefined behavior.
 *
 * <p><b>Sentinel values</b>: {@link #BEGINNING} (-1) and {@link #NOW} are protocol-defined
 * special values. All other offsets are opaque server-issued tokens.
 */
public final class Offset implements Comparable<Offset> {
    /** Start of stream. Equivalent to omitting offset parameter. */
    public static final Offset BEGINNING = new Offset("-1");

    /** Current tail position. Skips existing data, reads only future appends. */
    public static final Offset NOW = new Offset("now");

    private final String value;

    public Offset(String value) {
        this.value = Objects.requireNonNull(value, "offset value cannot be null");
    }

    public static Offset of(String value) {
        if ("-1".equals(value)) return BEGINNING;
        if ("now".equals(value)) return NOW;
        return new Offset(value);
    }

    public String getValue() {
        return value;
    }

    /**
     * Lexicographic comparison. Valid ONLY for offsets from the same stream.
     */
    @Override
    public int compareTo(Offset other) {
        return this.value.compareTo(other.value);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof Offset)) return false;
        return value.equals(((Offset) o).value);
    }

    @Override
    public int hashCode() {
        return value.hashCode();
    }

    @Override
    public String toString() {
        return value;
    }
}

File: packages/durable-streams/packages/client-dotnet/src/DurableStreams/Types/Offset.cs
namespace DurableStreams;

/// <summary>
/// Opaque stream offset token.
///
/// Per protocol specification (Section 6):
/// - Offsets are OPAQUE: do not parse or construct arbitrary offset values
/// - Offsets are LEXICOGRAPHICALLY SORTABLE: comparison operators are valid
///   and reflect stream position ordering
/// - Only use offsets received from the server (Stream-Next-Offset header,
///   control events) or the sentinel values (Beginning, Now)
/// </summary>
public readonly struct Offset : IEquatable<Offset>, IComparable<Offset>
{
    /// <summary>
    /// Beginning of stream (offset -1).
    /// </summary>
    public static readonly Offset Beginning = new("-1");

    /// <summary>
    /// Current tail position (skip historical data).
    /// </summary>
    public static readonly Offset Now = new("now");

    private readonly string _value;

    /// <summary>
    /// Create an offset from a string value.
    /// </summary>
    public Offset(string value)
    {
        _value = value ?? throw new ArgumentNullException(nameof(value));
    }

    /// <summary>
    /// Returns the raw offset string.
    /// </summary>
    public override string ToString() => _value;

    /// <summary>
    /// Returns true if this is the Beginning offset (-1).
    /// </summary>
    public bool IsBeginning => _value == "-1";

    /// <summary>
    /// Returns true if this is the Now offset.
    /// </summary>
    public bool IsNow => _value == "now";

    /// <summary>
    /// Returns true if this offset has a valid value.
    /// </summary>
    public bool HasValue => !string.IsNullOrEmpty(_value);

    /// <summary>
    /// Implicit conversion from string.
    /// </summary>
    public static implicit operator string(Offset offset) => offset._value;

    /// <summary>
    /// Explicit conversion from string to Offset.
    /// Use this for server-provided offset values. For sentinel values, prefer Offset.Beginning or Offset.Now.
    /// </summary>
    public static explicit operator Offset(string value) => new(value);

    /// <summary>
    /// Lexicographic comparison (per protocol specification).
    /// </summary>
    public int CompareTo(Offset other) =>
        string.Compare(_value, other._value, StringComparison.Ordinal);

    /// <inheritdoc />
    public bool Equals(Offset other) => _value == other._value;

    /// <inheritdoc />
    public override bool Equals(object? obj) => obj is Offset o && Equals(o);

    /// <inheritdoc />
    public override int GetHashCode() => _value?.GetHashCode() ?? 0;

    /// <summary>Equality operator.</summary>
    public static bool operator ==(Offset left, Offset right) => left.Equals(right);

    /// <summary>Inequality operator.</summary>
    public static bool operator !=(Offset left, Offset right) => !left.Equals(right);

    /// <summary>Less than operator (lexicographic).</summary>
    public static bool operator <(Offset left, Offset right) => left.CompareTo(right) < 0;

    /// <summary>Greater than operator (lexicographic).</summary>
    public static bool operator >(Offset left, Offset right) => left.CompareTo(right) > 0;

    /// <summary>Less than or equal operator.</summary>
    public static bool operator <=(Offset left, Offset right) => left.CompareTo(right) <= 0;

    /// <summary>Greater than or equal operator.</summary>
    public static bool operator >=(Offset left, Offset right) => left.CompareTo(right) >= 0;
}
